src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs

```diff
@@ -146,13 +145,83 @@ private static bool ApplyChangesFromTC31(JObject sarifLog)
             return modifiedLog;
         }
 
+        private static bool MoveRulesMetadataAndConfiguration(JObject run)
+        {
+            if (!(run["resources"] is JObject resources))
+            {
+                return false;
+            }
+
+            JObject tool = (JObject)run["tool"];
+
+            // https://github.com/oasis-tcs/sarif-spec/issues/311
+
+            // 1. 'run.resources.messageStrings' moves to 'run.tool.globalMessageStrings'
+            if (resources["messageStrings"] is JObject messageStrings)
+            {
+                tool["globalMessageStrings"] = messageStrings;
+                resources["messageStrings"] = null;
+            }
+
+            // 2. 'run.resources.rules' moves to 'run.tool.rulesMetadata'
+            if (resources["rules"] is JArray rules)
+            {
+                foreach(JObject rule in rules)
+                {
+                    RenameProperty(rule, previousName: "configuration", newName: "defaultConfiguration");
+
+                    if (rule["defaultConfiguration"] is JObject ruleConfiguration)
+                    {
+                        RenameProperty(ruleConfiguration, previousName: "defaultLevel", newName: "level");
+                        RenameProperty(ruleConfiguration, previousName: "defaultRank", newName: "rank");
+                    }
+                }
+                tool["rulesMetadata"] = rules;
+            }
+
+            // 3. We do not need any accommodation for the addition of 
+            // 'tool.notificationsMetadata', as this did not exist previously
+
+            // 4. Zap 'rules.resources' entirely
+            run["resources"] = null;
+
+            return true;
+        }
+
+        private static bool UpdateBaselineExistingStateToUnchanged(JObject result)
+        {
+            bool modifiedLog = false;
+
+            // Rename 'existing' baseline state to 'unchanged'
+            // (as part of adding the 'updated' state, which 
+            // will not exist in any legacy SARIF logs.
```

Comment:
>. [](start = 54, length = 1)

Close paren. And again, I think this and the issue URI should be a method level comment.